library/std/src/sys/unix/fd.rs

```diff
@@ -66,6 +69,10 @@ const fn max_iov() -> usize {
     16 // The minimum value required by POSIX.
 }
 
+// ESP-IDF does not (yet) support duplication of file descriptors, so we are emulating it here
```

Comment:
That kind of emulation seems dicy since duped file descriptors normally don't share a seek position, nonblocking status and various other properties. Here they would.

How much would break if `duplicate` always returned an error?

Reply:
This is here just to support off-the-shelf Rust crates (I've so far only found [Rocket](https://rocket.rs/) relying on it, but I'm unsure how many more are there). But no pressing use case, and Rocket is honestly a bit too big for this MCU anyway.

In any case and based on your explanations, this implementation is as flawed as it gets, so I'll just eradicate it completely tomorrow.

If proper file duplication semantics turns out to be necessary, I don't see any other way besides implementing it in the ESP-IDF framework itself.

Reply:
Hey I was [refreshing my knowledge](https://www.gnu.org/software/libc/manual/html_node/Duplicating-Descriptors.html) on `dup()`/`dup2()`/`FD_DUPFD` this morning (as I wrote this code 9 months ago!), and I don't think your interpretation "since duped file descriptors normally don't share a seek position, nonblocking status and various other properties" is correct.

According to the above link, it is exactly the opposite: dup-ed file descriptors DO share seek position, nonblocking status and various other properties. Duped file descriptors are nothing else but aliases of each other, and their only (limited) use seems to be lifetime management, where you can close the original file descriptor while the duped one continues to live and vice versa.

I think more or less this is what my implementation also does, doesn't it? (we can forget about the `CLOEXEC` semantics as that's irrelevant when you don't have processes)?

As for the use case with Rocket (remembered that too) - Rocket was `dup`-ing the file descriptor (socket really) so that it spins a separate socket writer from socket reader, and manages their lifetimes separately too.

Reply:
Yeah, you're right. I guess I was thinking of the mechanism that reopens a file via procfs, which isn't relevant here.

Ignore everything I have said here. 😔